src/pyhf/constraints.py

```diff
@@ -38,85 +44,105 @@ def _precompute(self):
+
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _precompute(self):
+        if not self.parameter_helper.index_selection:
+            return
+        tensorlib, _ = get_backend()
+        self.batched_sigmas = tensorlib.astensor(self._batched_sigmas)
+        self.normal_data = tensorlib.astensor(self._normal_data, dtype='int')
+        self.access_field = tensorlib.astensor(self._access_field, dtype='int')
 
     def logpdf(self, auxdata, pars):
-        if self.normal_data is None:
-            return 0
         tensorlib, _ = get_backend()
+        if not self.parameter_helper.index_selection:
+            return (
+                tensorlib.zeros(self.batch_size)
+                if self.batch_size is not None
+                else tensorlib.astensor(0.0)[0]
+            )
+
+        pars = tensorlib.astensor(pars)
+        if self.batch_size == 1 or self.batch_size is None:
+            batched_pars = tensorlib.reshape(
+                pars, (self.batch_size or 1,) + tensorlib.shape(pars)
+            )
+        else:
+            batched_pars = pars
+
+        flat_pars = tensorlib.reshape(batched_pars, (-1,))
+        normal_means = tensorlib.gather(flat_pars, self.access_field)
+
+        # pdf pars are done, now get data and compute
+        auxdata = tensorlib.astensor(auxdata)
         normal_data = tensorlib.gather(auxdata, self.normal_data)
-        normal_means = tensorlib.gather(pars, self.normal_mean_idc)
-        normal = tensorlib.normal_logpdf(normal_data, normal_means, self.normal_sigmas)
-        return tensorlib.sum(normal)
+        normal = tensorlib.normal_logpdf(normal_data, normal_means, self.batched_sigmas)
+        result = tensorlib.sum(normal, axis=1)
+        if self.batch_size is None:
+            return result[0]
+        return result
 
 
 class poisson_constraint_combined(object):
-    def __init__(self, pdfconfig):
+    def __init__(self, pdfconfig, batch_size=None):
+        self.batch_size = batch_size
         # iterate over all constraints order doesn't matter....
 
         self.par_indices = list(range(len(pdfconfig.suggested_init())))
         self.data_indices = list(range(len(pdfconfig.auxdata)))
-        self.mod_and_slice = [
-            (pdfconfig.param_set(cname), pdfconfig.par_slice(cname))
+        self.parset_and_slice = [
+            pdfconfig.param_set(cname) for cname in pdfconfig.auxdata_order
+        ]
+
+        pnames = [
+            cname
             for cname in pdfconfig.auxdata_order
+            if pdfconfig.param_set(cname).pdf_type == 'poisson'
         ]
-        self._precompute()
-        events.subscribe('tensorlib_changed')(self._precompute)
 
-    def _precompute(self):
-        tensorlib, _ = get_backend()
+        parfield_shape = (self.batch_size or 1, len(pdfconfig.suggested_init()))
+        self.parameter_helper = ParamViewer(parfield_shape, pdfconfig.par_map, pnames)
```

Comment:
call it `self.paramviewer` instead of yet another name.

Reply:
done